### PR TITLE
Dynamic: support open enums

### LINKED ISSUE #1243: Open enums: Dynamic support
In #1137, open enum support was added. Unfortunately, the `DynamicModelCompiler` treats all enums as closed. I don't think this can be worked around in userland, because ultimately you end up having to go from `String` / `Int` to `A`, which is abstract to you.

I propose that we add a condition for `@openEnum` just like in the codegen, and produce open enum tags in these cases.

### PR DESCRIPTION
**Pull request by kubukoz**
Closes #1243

### CODE REVIEW

File: modules/dynamic/src/smithy4s/dynamic/internals/DynamicModelCompiler.scala [hunk 1]
```diff
@@ -256,9 +256,33 @@ private[dynamic] object Compiler {
             )
           }
 
-      val theEnum = stringEnumeration(values, values)
+      if (shape.traits.contains(IdRef("alloy#openEnum"))) {
+        // the runtime representation of normal enums is Int, but for open enums it's String to support arbitrary unknown values.
```
Comment (kubukoz) [inline modules/dynamic/src/smithy4s/dynamic/internals/DynamicModelCompiler.scala:255]: **question:** should we go for a `String` representation for all the cases?

File: modules/dynamic/src/smithy4s/dynamic/internals/DynamicModelCompiler.scala [hunk 2]
```diff
@@ -289,14 +313,34 @@ private[dynamic] object Compiler {
...
-        theEnum
-      )
+        updateWithHints(
+          id, {
+            allHints(shape.traits)
+          },
+          theEnum
+        )
+      }
     }
```
Comment (kubukoz) [inline modules/dynamic/src/smithy4s/dynamic/internals/DynamicModelCompiler.scala:332]: **issue:** pending - support for 1.x enums. I'll add those later on  in this PR.

Comment (kubukoz) [reply modules/dynamic/src/smithy4s/dynamic/internals/DynamicModelCompiler.scala:332]: added now

File: modules/dynamic/src/smithy4s/dynamic/internals/DynamicModelCompiler.scala (hunk 1 above)
Comment (Baccata) [reply modules/dynamic/src/smithy4s/dynamic/internals/DynamicModelCompiler.scala:255]: I suppose that'd make sense.

Comment (Baccata) [reply modules/dynamic/src/smithy4s/dynamic/internals/DynamicModelCompiler.scala:255]: Although, int is fine for the closed-enum case

**Review by Baccata — APPROVED** [review]

**Review by Baccata — APPROVED** [review]

**Pull request merged**